sgkit/stats/pedigree.py

```diff
@@ -165,3 +170,524 @@ def topological_argsort(parent: ArrayLike) -> ArrayLike:  # pragma: no cover
+    return True
+
+
+@njit(cache=True)
+def _raise_on_half_founder(
+    parent: ArrayLike, tau: ArrayLike = None
+) -> None:  # pragma: no cover
+    for i in range(len(parent)):
+        p = parent[i, 0]
+        q = parent[i, 1]
+        if tau is None:
+            tau_p = 1
+            tau_q = 1
+        else:
+            tau_p = tau[i, 0]
+            tau_q = tau[i, 1]
+        if (p < 0 and q >= 0) and tau_p > 0:
+            raise ValueError("Pedigree contains half-founders")
+        elif (q < 0 and p >= 0) and tau_q > 0:
+            raise ValueError("Pedigree contains half-founders")
+    return
+
+
+@njit(cache=True)
+def _diploid_self_kinship(
+    kinship: ArrayLike, i: int, p: int, q: int
+) -> None:  # pragma: no cover
+    # self kinship of i with parents p and q
+    if (p < 0) or (q < 0):  # founder or half-founder
+        kinship[i, i] = 0.5
+    else:  # non-founder
+        kinship[i, i] = (1 + kinship[p, q]) / 2
+    return
+
+
+@njit(cache=True)
+def _diploid_pair_kinship(
+    kinship: ArrayLike, i: int, p: int, q: int, j: int
+) -> None:  # pragma: no cover
+    # kinship of i with j where j < i and i has parents p and q
+    kinship_pj = kinship[p, j] if p >= 0 else 0
+    kinship_qj = kinship[q, j] if q >= 0 else 0
+    kinship_ij = (kinship_pj + kinship_qj) / 2
+    kinship[i, j] = kinship_ij
+    kinship[j, i] = kinship_ij
+    return
+
+
+@njit(cache=True, parallel=True)
+def kinship_diploid(
+    parent: ArrayLike, allow_half_founders: bool = False, dtype: type = np.float64
+) -> ArrayLike:  # pragma: no cover
+    """Calculate pairwise expected kinship from a pedigree assuming all
+    individuals are diploids.
+
+    Parameters
+    ----------
+    parent
+        A matrix of shape (samples, parents) containing the indices of each
+        sample's parents with negative values indicating unknown parents as
+        defined in :data:`sgkit.variables.parent_spec`.
+
+    allow_half_founders
+        If False (the default) then a ValueError will be raised if any
+        individuals only have a single recorded parent.
+        If True then the unrecorded parent will be assumed to be
+        a unique founder unrelated to all other founders.
+    dtype
+        The dtype of the returned matrix.
+
+    Returns
+    -------
+    kinship
+        A square matrix of kinship estimates with self-kinship values on
+        the diagonal.
+
+    Raises
+    ------
+    ValueError
+        If the pedigree contains a directed loop.
+    ValueError
+        If the pedigree contains half-founders and allow_half_founders=False.
+    ValueError
+        If the parents dimension does not have a length of 2.
+    """
+    if parent.shape[1] != 2:
+        raise ValueError("Parent matrix must have shape (samples, 2)")
+    if not allow_half_founders:
+        _raise_on_half_founder(parent)
+    n = len(parent)
+    kinship = np.empty((n, n), dtype=dtype)
+    ordered = _is_pedigree_sorted(parent)
+    if ordered:
```

Comment:
Wouldn't it be simpler to require that the dataset be ordered, or to have the argsorting stored in the dataset? I worry that (a) we're duplicating code here and (b) that people will end up calling ``topological_argsort`` many times when running different algorithms on the pedigree.

It wouldn't be so bad to raise an error if the sorting wasn't present, would it?

Reply:
I did consider requiring sorted data but that requires sorting the full dataset by the samples dimension (including all genotype data and etc.) which isn't desirable. Storing the sort order is an option but it's a bit odd because the ith element in the sort order doesn't correspond to the ith sample, so it doesn't really fit with the xarray paradigm. `topological_argsort`is actually quite cheap and is trivial compared to the kinship matrix (O(N^2)).

Reply:
Yeah, good point. Could we avoid the code duplication then by having the argsort = np.arange(...) in the ordered case? I doubt it would make much perf difference (but maybe I'm wrong?)